src/nupic/engine/Link.cpp

```diff
@@ -396,21 +556,49 @@ void Link::write(LinkProto::Builder& proto) const
   proto.setSrcOutput(srcOutputName_.c_str());
   proto.setDestRegion(destRegionName_.c_str());
   proto.setDestInput(destInputName_.c_str());
+  proto.setOutputElementType(src_->getData().getType());
+  proto.setOutputElementCount(src_->getData().getCount());
+
+  // Save delayed outputs
+  auto delayedOutputsBuilder = proto.initDelayedOutputs(propagationDelay_);
+  for (size_t i=0; i < propagationDelay_; ++i)
+  {
+    _copyArrayToArrayProto(srcBuffer_[i], delayedOutputsBuilder[i]);
+  }
 }
 
+
 void Link::read(LinkProto::Reader& proto)
 {
+  const auto delayedOutputsReader = proto.getDelayedOutputs();
+
   commonConstructorInit_(
       proto.getType().cStr(), proto.getParams().cStr(),
       proto.getSrcRegion().cStr(), proto.getDestRegion().cStr(),
       proto.getSrcOutput().cStr(), proto.getDestInput().cStr(),
-      0/*propagationDelay ZZZ TODO get from proto*/);
+      delayedOutputsReader.size()/*propagationDelay*/);
+
+  // Initialize the propagation delay buffer
+  initPropagationDelayBuffer_(propagationDelay_,
+                              (NTA_BasicType)proto.getOutputElementType(),
+                              proto.getOutputElementCount());
+
+  // Populate delayed outputs
+
+  for (size_t i=0; i < propagationDelay_; ++i)
+  {
+    _copyArrayProtoToArray(delayedOutputsReader[i], srcBuffer_[i]);
+  }
 }
 
+
 namespace nupic
 {
   std::ostream& operator<<(std::ostream& f, const Link& link)
   {
+    // TODO Shouldn't this be implemented in terms of Link::toString? (assuming
```

Comment:
@scottpurdy, I was thinking that `operator<<` could call `toString` to generate the content, then dump the string to the stream. If `toString` is for debugging, why does it render only a subset of the data versus `operator<<`? If `toString` rendered the same info that `operator<<` does right now, then the logic could be shared between the two via `toString`. Wouldn't it be nice for consistency?